# Simple Twitter Analytics

Aim: To write simple python scripts to collect data from Twitter using a python wrapper of Twitter API - tweepy. Then fit the InterTweet Interval (ITI) into a power distribution, to check whether or not it follows a powerlaw (for reference of Powerlaw: http://tuvalu.santafe.edu/~aaronc/powerlaws/).




## Collecting Twitter Data Using Twitter API

Collect a certain number of Tweets by searching a hashtag or sent within a bounding box using Twitter Stream API, and store all the collected tweets into a JSON file.

##### Introduction to the Twitter Stream API

Twitter provides two kinds of API for developers, the REST APIs and Streaming APIs. The REST APIs provides programmatic access to read and write Twitter data. While the Streaming APIs continuously deliver new responses to REST API queries over a long-lived HTTP connection. It receives updates on the latest Tweets matching a search query, stay in sync with user profile updates, and more. In this lab, since the target is to study the inter-tweet disctibution, we need to collect real-time tweets. Thus, we will use the Stream API based on a Python implementation "tweepy". For more information about Twitter API, please read the information on https://dev.twitter.com/overview/documentation 

##### Register: Twitter Application Managment https://apps.twitter.com

The Twitter APIs require to authorized request to Twitter via an OAuth endpoints. To start,
- Login to your Twitter account
- Go to https://apps.twitter.com/app/ 
- Create a new app, enter a valid app name and URL
- Once done, you will go to another page.
- First switch to the permission tab and change the permissions to read write and direct messages.
- Then switch to the Keys and Access Tokens tab and generate the consumer key and secret, as well as the access token and secret (By selecting 'Create my access token' if needed).

You are now ready to proceed.



##### Using and Understanding JSON format

All the Twitter API search results are returned in a JSON format. JSON is a way to encode complicated information in a platform-independent way. It could be considered the lingua franca of information exchange on the Internet. It is a rather simplistic and elegant way to encode complex data structures. To encode and decode JSON, we need the package "json". More about JSON: https://docs.python.org/2/library/json.html

### Task #1

Start a Twitter Stream API implemented by the Python package "tweepy" to collect a certain number of tweets of a specific location. (pip install tweepy if needed)

Using the following python code collect 100 tweets of a location of your choice. 

In [1]:
import sys
sys.path.insert(0, 'C:\\Users\\rchanumo\\Downloads\\lab6\\lab6')
import tweepy
import json, time
from IPython.display import clear_output
from tweepy.streaming import StreamListener
from tweepy import OAuthHandler
from tweepy import Stream

In [2]:
# Go to http://dev.twitter.com and create an app.
# The consumer key and secret will be generated for you after
consumer_key="jPmqMdr53YRCzTcOO78pUXcOH"
consumer_secret="44RY6JqCfCs9jU2lpPLPLinVl4AzOCxar1VkFWBjXxiCtz1PGS"

# After the step above, you will be redirected to your app's page.
# Create an access token under the the "Your access token" section
access_token="3624684564-95V17JDIbJTZ0LwFjX6kkWWoQMspzuZRZryrMEq"
access_token_secret="zmdWpt8VdU6bKo53S51Mbu6NM0nQo0q9BpPefxVBrlOLb"

# List of cities: the numbers are the centers of each city, with square -/+ 0.2, using their latitudes and
# longitudes. Feel free to add your hometown and analyze if you want!

location_list = {'New York City, New York' : [-74.00 - 0.2, 40.71 - 0.2, -74.00 + 0.2, 40.71 + 0.2],
                 'Chicago, Illinois' : [-87.63 - 0.2, 41.88 - 0.2, -87.63 + 0.2, 41.88 + 0.2],
                 'Bloomington, Indiana' : [-86.53 - 0.2, 39.16 - 0.2, -86.53 + 0.2, 39.16 + 0.2], 
                 'Iowa City, Iowa' : [-91.53 - 0.2, 41.47 - 0.2, -91.53 + 0.2, 41.47 + 0.2], 
                 'Ann Arbor, Michigan' : [-83.75 - 0.2, 42.28 - 0.2, -83.75 + 0.2, 42.28 + 0.2],
                 'East Lansing, Michigan' :[-84.48 - 0.2, 42.73 - 0.2, -84.48 + 0.2, 42.73 + 0.2],
                 'Lincoln, Nebraska' : [-96.68 - 0.2, 40.81 - 0.2, -96.68 + 0.2, 40.81 + 0.2], 
                 'Columbus, Ohio' : [-82.98 - 0.2, 39.98 - 0.2, -82.98 + 0.2, 39.98 + 0.2], 
                 'State College, Pennsylvania' : [-77.86 - 0.2, 40.79 - 0.2, -77.86 + 0.2, 40.79 + 0.2], 
                 'West Lafayette, Indiana' : [-86.91 - 0.2, 40.44 - 0.2, -86.91 + 0.2, 40.44 + 0.2], 
                 'Madison, Wisconsin' : [-89.4 - 0.2, 43.07 - 0.2, -89.4 + 0.2, 43.07 + 0.2], 
                 'Champaign, Illinois' : [-88.27 - 0.2, 40.12 - 0.2, -88.27 + 0.2, 40.12 + 0.2]
                }

# The JSON listener class.

class JSONListener(StreamListener):
    """ A listener handles tweets are the received from the stream.
    This is a basic listener that just prints received tweets to stdout.

    """
    def __init__(self, fprefix = 'NY100', max_tweets = 100):
        self.counter = 0
        self.fprefix = fprefix
        self.output  = open('./data/' + fprefix + '.json', 'w')
        self.max_tweets = max_tweets
        
    def on_data(self, data):
        if self.counter < self.max_tweets:
            self.output.write(data)
            self.counter += 1
          
            clear_output()
            print( 'Tweets collected so far: {}'.format(self.counter))
            # Print the information
            # Comment these lines of code when you are crawling data for assignment 3!!!
            #'''
        #    status = json.loads(data)
        #    print 'Tweet No. {} at [Time = {}]'.format(self.counter, status['created_at'])
        #    print status['text']
            #'''
            return True
        else:
            print ('Totally {} tweets collected'.format(self.max_tweets))
            return False
        

    def on_error(self, status):
        print ('Error: ' + str(status))


In [3]:
if __name__ == '__main__':
    auth = OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_token_secret)
    
    listen = JSONListener(fprefix = "NY100", max_tweets = 100)
    city = location_list['New York City, New York']
    
    stream = Stream(auth, listen)
    stream.filter(locations=city)
    
    print ('Job completed. Disconnecting stream...')
    stream.disconnect()


Tweets collected so far: 100
Totally 100 tweets collected
Job completed. Disconnecting stream...


## Tweets Analytics

Fit a power distribution of the InterTweet Interval based on 

   1. The data collected in Exercise 1 for a location of your choice (100 tweets)
   2. A given dataset in ./data/CU.json (This corresponds to 1000 tweets from Champaign) 

To facillitate the powerlaw calculation, a Python package "powerlaw" will be used.



## Simple Analytics of ITI (InterTweet Interval) based on the data collected 

### Data preprocessing:

Import JSON file, and read the information to memory. For this, use ['timestamp_ms'] which is measured in ms. first convert the time into seconds and plot the time series distribution. Using plt.hist() and plt.show() functions to plot the distribution. And the number of bins could be set as the count of tweets.

In [3]:
#store the time series in an array/list, named 'data'
import matplotlib.pyplot as plt
import sys, json
import numpy as np
#Don't use inline option if you don't want to show figure in notebook.
#%matplotlib inline



In [4]:
data = []
fp = open('./data/NY100.json','r')
#traverse through tweets in file and find timestamps
for line in fp:
    #print(type(line))
    if line not in ['\n' , '\r\n']:
        tweet_as_dictionary = json.loads(line)
        timestamp = int(tweet_as_dictionary['timestamp_ms'])/1000
        data.append(timestamp)
#print(len(data))
plt.hist(data, bins = 100)
plt.title('Time Series Distribution')
plt.savefig('Timestamps_NY100.png')
plt.close()

Calculate Inter-Tweet Intervals (ITI). Note, by turning the time from ms to sec, the time series turns to be a float number series, which is a continuous distribution.  Sort the data. Then calculate inter-tweet intervals.



In [5]:
import numpy as np
#convert list to np array
data = np.array(data) 
#sort timestamps and find ITI by finding differnce between consecutive timestamps
x = np.diff(np.sort(data))

#plot ITI
plt.hist(np.sort(x), bins = 100)
plt.title('ITI')
plt.savefig('ITI_NY100.png')
plt.close()

### Fit the Distribution of ITI using Power-Law. 

A continuous power-law distribution is one described by a probability density function $p(x)$ such that $p(x)=C f(x)$, where $C$ is a normalization constant, $X$ is the observd value, and $f(x) = x^{-\alpha}$.

Typically, the explicit form of a power-law distribution is:
\begin{equation} p(x) = \frac{\alpha-1}{x_{min}} (\frac{x}{x_{min}})^{-\alpha} \end{equation}
where $x_{min}$ is the lower bound to the power-law distribution such that it holds $\forall x \geq x_{min}$ with $\alpha \geq 1$.

Sometimes, it is much easier to consider the (complementary) cumulative distribution function (CCDF) of a power-law distributed variable, which is denoted as $\bar{P}(x)=Pr(X\geq x)$. In the continuous case, 
\begin{equation}
\bar{P}(x)=\int_x p(x')dx' = (\frac{x}{x_{min}})^{-\alpha + 1}
\end{equation}

Note that the cumulative distribution function (CDF) is then simply 
\begin{equation}
P(x) = 1 - \bar{P}(x)
\end{equation}

#### To estimate the parameters in the power-law, you need to implement a function to calculate the estimator of $\alpha$, in the continuous case, as 
\begin{equation}
\hat{\alpha}=1+n(\sum_{i=1}^n \ln \frac{x_i}{x_{min}})^{-1}
\end{equation}

\begin{equation}
x_i \geq x_{min}, n = \text{len}(x_i>=x_{min})
\end{equation}

 

In this implementation, you need to assume the parameter $x_{min}$ is given or by default is 0.

In [6]:
# Implementation of estimator of alpha for power-law
import numpy 
import time
import pylab
from numpy import * 

In [7]:
def alpha(x, xmin = 0):
    # Your implementation here. 
    # Make sure to add a small value to the denominator of the log so that the term does
    # not blow up for small values of x_min
    atemp = 0
    z = x[x>=xmin]
    for xi in z:
        if xi != 0:
            #print(xi)
            atemp = atemp + log((xi + 0.5)/(xmin + 0.5))
            #print(atemp)
    if atemp != 0:
        a = 1 + (len(z)/atemp)
    else:
        a = 1
       
    return a

####  Estimating the lower bound on power-law behavior, $x_{min}$

The power-law fitting relies on the estimation of $x_{min}$. An efficient algorithm to estimate the lower bound is proposed by Clauset et al., and the fundamental idea behind it is: we choose the value of $\hat{x}_{min}$ that makes the probability distribution of the measured data and the best-fit power-law model as similar as possible above $x_{min}$ (i.e., $\forall x \geq x_{min}$).

To measure the distance between two probability distributions, the commonest distance used is Kolmogorov-Smirnov or KS statistic, which is simply the maximum distance between the CDFs of the data and the fitted model:
\begin{equation}
D=\max_{x\geq x_{min}} |S(x)-P(x)|
\end{equation}
and the best $x_{min}$ is achieved by
\begin{equation}
\hat{x}_{min} = \arg\min_{x_{min} \in X} D_{x_{min}}
\end{equation}

Here, a function to calculate the KS statistic is given, as kstest(xmin, x)

In [8]:
def kstest(xmin,x):
    x = x[x>=xmin]
    n = len(x)
    if n == 0: return numpy.inf
    # calculate the best power-law fit
    a = alpha(x, xmin)
    # Actual CDF
    cx = numpy.arange(n,dtype='float')/float(n)
    # CDF of the fitted power-law distribution
    cf = 1-(xmin/x)**(a-1)
    ks = max(abs(cf-cx))
    return ks

#### Use the above functions, to implement the power-law fitting function:

In [9]:
def plfit(x):
    # Your implementation here, ensure x is sorted
    '''
    Tips: choose x_min candidate from
    xmins,argxmins = numpy.unique(x,return_index=True)
    search the candidate xmins to determine the best xmin and alpha
    and use the function numpy.argmin() to select minimal value
    '''
   
    # The calculation of Likelihood is: L = n*log((alpha-1)/xmin) - alpha*sum(log(z/xmin)), with z = x[x>xmin]
    xmins,argxmins = numpy.unique(x,return_index=True);
    #initialize ks to a big value
    ks = float('inf')
    for xmin in xmins:
        if xmin != max(x):
            #for eash xmin find ks
            ks_temp = kstest(xmin,x)
            #print(ks_temp)
            if(ks > ks_temp):
                # find xmin which results in least ks
                ks = ks_temp
                x_min = xmin
    
    
    a = alpha(x,x_min)
    z = x[x>x_min]
    n = len(x)
    L = n*log((a-1)/x_min) - a*sum(log(z/x_min))
    # Return value is (xmin, a, ks, L)
    # a - alpha, ks - KS distance, L - likelihood
    return (x_min, a, ks, L)

### Test the Goodness of PowerLaw fitting

Given an observed data set and a hypothesized power-law distribution from which the data are drawn, we would like to know whether the hypothesis is a plausible one, given the data. A standard approach is to use a goodness-of-fit test, which generate a $p$-value that quantifies the plausibility of the hypothesis. 

The basic idea of such test is based on measurement of the "distance" between the distribution of empirical data and the hypothesized model. This distance is compared with distance measurements for comparable synthetic data sets drawn from the same model, and the p-value is defined to be the fraction of the synthetic distances that are larger than the empirical distance.

If $p$ is large (close to 1), then the difference between the empirical data and the model can be attributed to the statistical fluctuations alone; if it is small, the model is not a plausible fit to the data. Typically, we use $0.1$ as threshold, that is if $p \geq 0.1$, the model fitting is plausible, otherwise not.

In the following is a function to test the $p$-value, use the function to test your power-law fitting, where xmin, alpha, ks is the returned value of plfit for the empirical data. And returned value is p-value and the ks vector for the random generated synthetic data.

In [10]:
import numpy
import numpy.random as npr

def test_pl(x, xmin, alpha, ks, niter=1e2, print_timing=False):
    """
    Monte-Carlo test to determine whether distribution is consistent with a power law

    Runs through niter iterations of a sample size identical to the input sample size.

    Will randomly select values from the data < xmin.  The number of values selected will
    be chosen from a uniform random distribution with p(<xmin) = n(<xmin)/n.

    Once the sample is created, it is fit using above methods, then the best fit is used to
    compute a Kolmogorov-Smirnov statistic.  The KS stat distribution is compared to the 
    KS value for the fit to the actual data, and p = fraction of random ks values greater
    than the data ks value is computed.  If p<.1, the data may be inconsistent with a 
    powerlaw.  A data set of n(>xmin)>100 is required to distinguish a PL from an exponential,
    and n(>xmin)>~300 is required to distinguish a log-normal distribution from a PL.
    For more details, see figure 4.1 and section

    **WARNING** This can take a very long time to run!  Execution time scales as 
    niter * setsize

    """
    niter = int(niter)

    ntail = sum(x >= xmin)
    ntot = len(x)
    nnot = ntot-ntail              # n(<xmin)
    pnot = nnot/float(ntot)        # p(<xmin)
    nonpldata = x[x<xmin]
    nrandnot = sum( npr.rand(ntot) < pnot ) # randomly choose how many to sample from <xmin
    nrandtail = ntot - nrandnot         # and the rest will be sampled from the powerlaw

    ksv = []
    if print_timing: deltat = []
    for i in range(niter):
        # first, randomly sample from power law
        # with caveat!  
        nonplind = numpy.floor(npr.rand(nrandnot)*nnot).astype('int')
        fakenonpl = nonpldata[nonplind]
        randarr = npr.rand(nrandtail)
        fakepl = randarr**(1/(1-alpha)) * xmin 
        fakedata = numpy.concatenate([fakenonpl,fakepl])
        if print_timing: t0 = time.time()
        # second, fit to powerlaw
        TEST = plfit(fakedata)
        # Get the ks of the fakedata fitting
        ksv.append(TEST[2])
        # Print time information if needed
        if print_timing: 
            deltat.append( time.time() - t0 )
            print ("Iteration %i: %g seconds" % (i, deltat[-1]))
    
    ksv = numpy.array(ksv)
    p = (ksv>ks).sum() / float(niter)

    print ("p(%i) = %0.3f" % (niter,p))
    
    if print_timing: print ("Iteration timing: %g +/- %g" % (numpy.mean(deltat),numpy.std(deltat)))

    return p, ksv

#### Perform the power-law fitting. Obtain its goodness-of-fit value.

In [11]:
#Output xmin, a, ks of best fit, and p-value.
(xmin, a, ks, L) = plfit(x) 
(p , ksv) = test_pl(x, xmin, a, ks)
print(xmin)
print(a)
print(ks)
print(p)

C:\Program Files\Anaconda3\lib\site-packages\ipykernel\__main__.py:10: RuntimeWarning: invalid value encountered in true_divide


p(100) = 0.000
3.27799987793
7.99236083177
0.895691729171
0.0


#### Fill out your answers here,  for the location of your choice. 
xmin = 3.27799987793

a = 7.99236083177

ks_value = 0.895691729171

p_value = 0.01

#### Plotting: Plot the CDF of the input data and the fitted power-law distribution, in a log-log chart.

Use the following code to plot the CDFs.

In [12]:
import matplotlib.pyplot as plt

def plotcdf(x, xmin, alpha, pointcolor='k', pointmarker='+', filename=''):
    """
    Plots CDF and powerlaw
    """
    # Generate CDF of x
    x=numpy.sort(x)
    n=len(x)
    xcdf = numpy.arange(n,0,-1,dtype='float')/float(n)
    
    q = x[x>=xmin]
    fcdf = (q/xmin)**(1-alpha)
    nc = xcdf[argmax(x>=xmin)]
    fcdf_norm = nc*fcdf

    # Draw the fitted distribution as a line
    D_location = argmax(xcdf[x>=xmin]-fcdf_norm)

    plt.clf()
    plt.cla()
    plt.loglog(x,xcdf,marker=pointmarker,color=pointcolor)
    plt.loglog(q,fcdf_norm,'r')

    if filename == '':
        plt.show()
    else:
        # plt.show()
        plt.savefig('3_'+filename) 
    plt.close()

In [13]:
# Call plotcdf() to plot the distribution here
# Remember to save your figure into file
# If filename = '' or no filename provided, it will be not be saved.

plotcdf(x, xmin, a, 'k', '+', filename='CDF_NY100')

#### About the fit

Do you think the fit is good? 

I dont think it is good fit because the p value=0.01 which is much less than 0.1
Also from the CDF plot showing the CDF of data after applying powerfit model(red line) and CDF of actual data( in blue) we see that they dont fit each other very well.

### Comparing with other models

Besides power-law, there are several other popular distributions widely used, such as exponential $f(x)=e^{-\lambda x}$, $C=\lambda e^{\lambda x_{min}}$, and log-normal distribution, etc.

We suggest to use an public available powerlaw fitting python package, the 'powerlaw' (which can be found in either pip or Canopy package manager). Also install mpmath if you are encountering errors.

Fit the data using exponential distribution, and compare with the power-law fitting by computing the likelihood ratio test. For more information about how to use powerlaw package, please see:

http://nbviewer.ipython.org/gist/anonymous/bb4e1dfafd9e90d5bc3d

http://pythonhosted.org/powerlaw/#powerlaw.Fit.distribution_compare

###### Hints:
1.  Use fit.distribution_compare() function to compare fittings between power_law and lognormal distribution. The output will be which kind of distribution fits the data better
2. Use plot_ccdf() function to draw the fitting. 
   * use fit.plot_ccdf() to plot the empiricial data distribution
   * use fit.power_law.plot_ccdf() to plot the power_law fitting curve
   * use fit.lognormal.plot_ccdf() to plot the log fitting curve


In [ ]:

import matplotlib.pyplot as plt
import powerlaw

fit = powerlaw.Fit(x, discrete=False)
# Using the functions 
# 1. fit.distribution_compare() to compare 
log_likelihood= fit.distribution_compare('power_law', 'lognormal')[0]
print ('Log likelihood is '+ str(log_likelihood))

# plot empirical CCDF, power law CCDF and log normal CCDFs
fig = fit.plot_ccdf(color='b', linewidth=2 , label='Empirical Data')
fit.power_law.plot_ccdf(color='r', linewidth=2, ax=fig , label='Power law') 
fit.lognormal.plot_ccdf(color='g', linewidth=2, ax=fig , label='Log normal') 

fig.set_ylabel(r"$p(X\geq x)$")
fig.set_xlabel(r"ITI")
handles, labels = fig.get_legend_handles_labels()
fig.legend(handles, labels, loc=3)
plt.show()
plt.savefig('4_compare_plots.png')
plt.close()

#### Based on the plot you get above, which fit is better?
As x increases we see that the plot corresponding to power_law deviates more from the actual distribution as compared to log normal. therefore, The lognormal fit is better. 

#### Based on the log-likelihoods you get above, which fit is better?
Since the value of Log liklihood ratio is -2.63748431219 and therefore negative the second distrbution i.e log normal gives a better fit.

## Repeat the process for the data given in ./data/CU.json


The data contains 1,000 tweets sent in Champaign. 

In [ ]:
#Get timestamps from data
data = []
fp = open('./data/CU.json','r')
for line in fp:
    #print(type(line))
    if line not in ['\n' , '\r\n']:
        tweet_as_dictionary = json.loads(line)
        timestamp = int(tweet_as_dictionary['timestamp_ms'])/1000
        data.append(timestamp)

#find ITI
data = np.array(data) 
x = np.diff(np.sort(data))

#Output xmin, a, ks of best fit, and p-value.
(xmin, a, ks, L) = plfit(x) 
(p , ksv) = test_pl(x, xmin, a, ks)
print(xmin)
print(a)
print(ks)
print(p)



#### Fill out your answers here, for Champaign:
x_min =37.0099999905

a = 19.3879084278

ks_value = 0.368083264369
 
p_value = 0.38

#### About the fit, for Champaign

How does the fitted distribution compare to the original data?

Do you think the fit is good? Explain:
I think tha fit is good since the p-value(0.38) is greater than 0.1

## ITI Fitting Using the Best Model

Run Twitter Crawler to collect 3,000 tweets choosing a specific location in the city list during the following week, and fit the ITI (Inter-Tweets Interval) distribution in the best model. (Tip: Choose your locations based on the pace of life!)

In [24]:
#Get timestamps from data
data = []
fp = open('./data/NY3000.json','r')
for line in fp:
    #print(type(line))
    if line not in ['\n', '\r\n']:
        tweet_as_dictionary = json.loads(line)
        timestamp = int(tweet_as_dictionary['timestamp_ms'])/1000
        data.append(timestamp)

#find ITI
data = np.array(data) 
x = np.diff(np.sort(data))

#Output xmin, a, ks of best fit, and p-value.
(xmin, a, ks, L) = plfit(x) 
(p , ksv) = test_pl(x, xmin, a, ks)
print(xmin)
print(a)
print(ks)
print(p)

C:\Program Files\Anaconda3\lib\site-packages\ipykernel\__main__.py:10: RuntimeWarning: invalid value encountered in true_divide


p(100) = 0.950
5.48400020599
22.8930968218
0.264007466927
0.95


x_min =5.48400020599

a = 22.8930968218

ks_value = 0.264007466927
 
p_value =0.950

## Questions: 

   * What's the main difference between the fittings between the data you collected (100 tweets), the 1,000 tweets, and the 3,000 tweets?
   
   The main difference is the p value. we seet that as the number of tweets collected increases the p value increases thus allowing the power law model to fit the data better.
   
   
   * Is a low p-value more desirable than a high p-value? 
   
   P values evaluate how well the sample data supports the idea that the null hypothesis is true. Thus, measures how compatible our data are with the null hypothesis. lower p value rejects the null hypothesis an higher p values support null hypothesis. since we are trying to fit the model to the emperical data, higher p value is prefered over lower p value. This is bacause lower p value rejects null hupothesis which menas there is diffrence between the data distribution of the fiited model and emperical data on the other hand higher p value(>0.1) data are likely with a true null threfore it is plausible to fit the data.